### ISSUE TITLE
Autoloading annotation classes

### ISSUE THREAD

**Issue by richard-ejem** [doctrine/annotations#103]
Why aren't annotation classes loaded by currently configured PHP autoload mechanisms?

It is called neither in https://github.com/doctrine/annotations/blob/master/lib/Doctrine/Common/Annotations/AnnotationRegistry.php#L123
nor in https://github.com/doctrine/annotations/blob/master/lib/Doctrine/Common/Annotations/DocParser.php#L457 (implicit autoload is explicitly turned off using class_exists($className, false) ).

It leads to this weird behavior:

```
spl_autoload_register(function() { eval('/** @Annotation */ class MyAnnotation {}'); });
$docParser = new DocParser();

// uncomment the following line to change the result
// class_exists('MyAnnotation');
try {
    $docParser->parse('/** @MyAnnotation */');
    echo "MyAnnotation exists";
} catch (AnnotationException $ex) {
    echo "MyAnnotation does not exist";
}
```

My suggestion would be to let the DocParser try to autoload the class.

**Comment by stof** [comment]
they can be, with a simple trick:

```php
AnnotationRegistry::registerLoader('class_exists');
```

This works fine as long as your autoloaders are not breaking ``class_exists``, i.e. they are not throwing an error when they fail to find a class.

The AnnotationRegistry system exists precisely because several big frameworks were failing this requirement in their autoloader at the time Doctrine Common 2.1 was written (including Doctrine Common itself and Zend Framework 1).
In the composer era, this is not an issue anymore generally.

**Comment by richard-ejem** [comment]
@stof Yes as I explained in P.S., I figured out this solution too. It works, but is one of dozens-things  you have to figure out why they aren't working by default.
Wouldn't it be better to make calling the autoloader a default behavior and make use of AnnotationRegistry opt-in? Because in these days AFAIK major frameworks don't have this issue anymore (as it actually also breaks the class_exists() itself).

**Comment by stof** [comment]
@richard-ejem changing it is planned for Doctrine 3. Changing it in 2.x might cause issues for some cases.

**Comment by Ocramius** [comment]
Cross-linking https://github.com/doctrine/annotations/pull/61
